CLS & TBT fix

qiyundai opened the pull request:
I noticed that sometimes multifunction floating CTA's initial transition would cause a huge CLS penalty. This PR is to make the multifunction button not transition until the initial delay passes or the user has interacted with it. We should get a better CLS score.

Resolves: [MWPW-136879](https://jira.corp.adobe.com/browse/MWPW-136879)

Test URLs:
- Before: https://stage--express--adobecom.hlx.page/express/create/flyer?martech=off
- After: https://multi-cta-fix--express--adobecom.hlx.page/express/create/flyer?lighthouse=on

qiyundai commented:
Honestly I'm not so sure anymore if this fixes anything CLS. It does still make the loading of the multifunction button look & feel better for the users and has a refactoring that reduces the CSS loaded for the regular version of the floating button. So it's still worth the PR I think.

vhargrave commented:
linters or tests are failing here

vhargrave commented:
failing lint & tests here too

vhargrave commented on express/blocks/how-to-steps-carousel/how-to-steps-carousel.js:
```diff
@@ -247,6 +247,28 @@ function layerTemplateImage(canvas, ctx, templateImg) {
   });
 }
 
+function convertCanvasToBlob(canvas, width, height) {
```
I love the idea, but did this change actually show a measurable difference?

vhargrave replied:
ok, sounds good!

vhargrave reviewed (changes requested):
I'm a bit on the fence with this one @qiyundai .
I'm not getting reliable test results when comparing this PR vs main on pagespeed.web.dev . Would you be ok with discussing this on Monday?

qiyundai commented on express/blocks/cta-carousel/cta-carousel.js:
```diff
@@ -26,6 +26,12 @@ export function decorateTextWithTag(textSource, options = {}) {
   return text;
 }
 
+export const windowHelper = {
```
I absolutely agree.. there are actually 2 changes in this PR that are related to this issue. I fear that we'll have to test less because we can't be triggering the redirect in tests.

Besides this, the added code in carousel.js for isTestEnv, the gen-ai-cards block would also be affected.

qiyundai commented:
I removed the worker stuff and added a couple intersection observers to not let how to steps blocks init animation on page load. I've noticed that sometimes causing long tasks on CPU throttled tests. Do you think you can give this another look? Thanks! @vhargrave

vhargrave commented on express/blocks/search-marquee/search-marquee.js:
```diff
@@ -334,8 +334,8 @@ async function buildSearchDropdown(block) {
 
-function decorateLinkList(block) {
-  const carouselItemsWrapper = block.querySelector(':scope > div:nth-of-type(2)');
+async function decorateLinkList(block) {
```
probably want to await this as well?

vhargrave commented on express/blocks/shared/carousel.js:
```diff
@@ -177,3 +181,14 @@ export default async function buildCarousel(selector, parent, options = {}) {
+
+export default async function buildCarousel(selector, parent, options = {}) {
+  // Load CSS then build carousel
+  parent.style.visibility = 'hidden';
```
why are we making the carousel not visible? 
block sections do not display until they're fully loaded.

vhargrave commented on express/scripts/utils.js:
```diff
@@ -560,14 +560,14 @@ export function linkPicture($picture) {
-    $a.append(...$parent.children);
-    $parent.append($a);
-  }
+  if (!$a) return;
```
I don't think this guard is necessary. It's less bytes to just use braces

qiyundai commented on express/blocks/color-how-to-carousel/color-how-to-carousel.js:
```diff
@@ -222,5 +222,15 @@ export default async function decorate(block) {
+    observer.unobserve(block);
+  };
+
+  const colorHowToObserver = new IntersectionObserver(onIntersect, { threshold: 0 });
```
I'm a little worried about making the threshold = 1 as what if the user has a slightly smaller viewport than the full height of the block then the animation will never start?

vhargrave replied:
you're right! I thought threshold worked slightly differently

vhargrave commented on express/blocks/how-to-steps-carousel/how-to-steps-carousel.js:
```diff
@@ -190,11 +190,18 @@ function buildHowToStepsCarousel(section, picture, block, howToDocument, rows, h
+    observer.unobserve(block);
+  };
+
+  const howToStepsObserver = new IntersectionObserver(onIntersect, { threshold: 0 });
```
Fair!
